webapi/setaltsig_test.go

```diff
@@ -0,0 +1,236 @@
+	getRawTransactionErr error
+}
+
+func (n *testNode) CanTicketVote(_ *dcrdtypes.TxRawResult, _ string, _ *chaincfg.Params) (bool, error) {
+	return n.canTicketVote, n.canTicketVoteErr
+}
+
+func (n *testNode) GetRawTransaction(txHash string) (*dcrdtypes.TxRawResult, error) {
+	return nil, n.getRawTransactionErr
+}
+
+func TestSetAltSig(t *testing.T) {
+	const testAddr = "DsVoDXNQqyF3V83PJJ5zMdnB4pQuJHBAh15"
+	tests := []struct {
+		name                 string
+		vspClosed, clearHash bool
+		deformReq            int
+		addr                 string
+		getRawTransactionErr error
+		canTicketNotVote     bool
+		isExistingAltSig     bool
+		canTicketVoteErr     error
+		wantCode             int
+	}{{
+		name:     "ok",
+		addr:     testAddr,
+		wantCode: http.StatusOK,
+	}, {
+		name:      "vsp closed",
+		vspClosed: true,
+		wantCode:  http.StatusBadRequest,
+	}, {
+		name:      "bad request",
+		deformReq: 1,
+		wantCode:  http.StatusBadRequest,
+	}, {
+		name:     "bad addr",
+		addr:     "xxx",
+		wantCode: http.StatusBadRequest,
+	}, {
+		name:     "addr wrong type",
+		addr:     "DkM3ZigNyiwHrsXRjkDQ8t8tW6uKGW9g61qEkG3bMqQPQWYEf5X3J",
+		wantCode: http.StatusBadRequest,
+	}, {
+		name:                 "error getting raw tx from dcrd client",
+		addr:                 testAddr,
+		getRawTransactionErr: errors.New("get raw transaction error"),
+		wantCode:             http.StatusInternalServerError,
+	}, {
+		name:             "error getting can vote from dcrd client",
+		addr:             testAddr,
+		canTicketVoteErr: errors.New("can ticket vote error"),
+		wantCode:         http.StatusInternalServerError,
+	}, {
+		name:             "ticket can't vote",
+		addr:             testAddr,
+		canTicketNotVote: true,
+		wantCode:         http.StatusBadRequest,
+	}, {
+		name:             "hist at max",
+		addr:             testAddr,
+		isExistingAltSig: true,
+		wantCode:         http.StatusBadRequest,
+	}}
+
+	for _, test := range tests {
+		ticketHash := randString(64, hexCharset)
+		data := tAltSigData()
+		data.AltSigAddr = test.addr
+		if test.isExistingAltSig {
+			if err := db.InsertAltSig(ticketHash, data); err != nil {
+				t.Fatalf("%q: unable to insert ticket: %v", test.name, err)
+			}
+		}
+
+		cfg.VspClosed = test.vspClosed
+
+		tNode := &testNode{
+			canTicketVote:        !test.canTicketNotVote,
+			canTicketVoteErr:     test.canTicketVoteErr,
+			getRawTransactionErr: test.getRawTransactionErr,
+		}
+		w := httptest.NewRecorder()
+		c, r := gin.CreateTestContext(w)
+
+		req := &SetAltSigRequest{
+			Timestamp:     time.Now().Unix(),
+			TicketHash:    ticketHash,
+			TicketHex:     randString(504, hexCharset),
+			ParentHex:     randString(504, hexCharset),
+			AltSigAddress: test.addr,
+		}
+		b, err := json.Marshal(req)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		handle := func(c *gin.Context) {
+			c.Set("DcrdClient", tNode)
+			c.Set("RequestBytes", b[test.deformReq:])
+			c.Set("CommitmentAddress", randString(35, addrCharset))
```

Comment:
`"CommitmentAddress"` no longer required

Reply:
Removed in 3727325